<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="raido-bot">

    <material name="grey">
        <color rgba="0.38 0.38 0.38 1.0" />
    </material>
    <material name="black">
        <color rgba="0 0 0 1.0" />
    </material>

    <xacro:property name="wheel_radius" value="0.06" />
    <xacro:property name="wheel_length" value="0.05" />
    <xacro:property name="base_length" value="0.5" />
    <xacro:property name="base_width" value="0.35" />
    <xacro:property name="base_height" value="0.11" />
    <xacro:property name="joint_offset" value="0.1" />

    <link name="base_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
            <mass value="5.0" />
            <inertia ixx="0.043125" ixy="0.0" ixz="0.0" iyy="0.0890625" iyz="0.0" izz="0.1040625" />
        </inertial>
        <collision name="base_link_col">
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
            <geometry>
                <box size="${base_length} ${base_width} ${base_height}" />
            </geometry>
        </collision>
        <visual name="base_link_vis">
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
            <geometry>
                <mesh filename="package://raido-bot_description/meshes/base_link.stl" scale="1 1 1" />
            </geometry>
            <material name="grey" />
        </visual>
    </link>

    <xacro:macro name="wheel_link" params="name x y z">
        <link name="${name}">
            <inertial>
                <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
                <mass value="1.0" />
                <inertia ixx="0.0025" ixy="0.0" ixz="0.0" iyy="0.0025" iyz="0.0" izz="0.001" />
            </inertial>
            <collision name="${name}_col">
                <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
                <geometry>
                    <cylinder radius="${wheel_radius}" length="${wheel_length}" />
                </geometry>
            </collision>
            <visual name="${name}_vis">
                <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
                <geometry>
                    <mesh filename="package://raido-bot_description/meshes/wheel_link.stl"
                        scale="1 1 1" />
                </geometry>
                <material name="black" />
            </visual>
        </link>

        <joint name="${name}_joint" type="continuous">
            <origin xyz="${x} ${y} ${z}" rpy="0.0 0.0 0.0" />
            <parent link="base_link" />
            <child link="${name}" />
            <axis xyz="0 1 0" />
            <limit lower="0.0" upper="0.0" effort="1.0" velocity="1.0" />
        </joint>
    </xacro:macro>

    <xacro:wheel_link name="wheel_link_fr" x="${(base_length / 2) - joint_offset}"
        y="${(base_width / 2) + (wheel_length / 2)}" z="0.0" />

    <xacro:wheel_link name="wheel_link_br" x="-${(base_length / 2) - joint_offset}"
        y="${(base_width / 2) + (wheel_length / 2)}" z="0.0" />

    <xacro:wheel_link name="wheel_link_fl" x="${(base_length / 2) - joint_offset}"
        y="-${(base_width / 2) + (wheel_length / 2)}" z="0.0" />

    <xacro:wheel_link name="wheel_link_bl" x="-${(base_length / 2) - joint_offset}"
        y="-${(base_width / 2) + (wheel_length / 2)}" z="0.0" />

    <gazebo>
        <plugin
            filename="ignition-gazebo-diff-drive-system"
            name="gz::sim::systems::DiffDrive">
            <left_joint>wheel_link_fl_joint</left_joint>
            <right_joint>wheel_link_fr_joint</right_joint>
            <wheel_separation>0.35</wheel_separation>
            <wheel_radius>0.06</wheel_radius>
            <odom_publish_frequency>0.1</odom_publish_frequency>
            <max_linear_acceleration>1</max_linear_acceleration>
            <min_linear_acceleration>-1</min_linear_acceleration>
            <max_angular_acceleration>2</max_angular_acceleration>
            <min_angular_acceleration>-2</min_angular_acceleration>
            <max_linear_velocity>0.5</max_linear_velocity>
            <min_linear_velocity>-0.5</min_linear_velocity>
            <max_angular_velocity>1</max_angular_velocity>
            <min_angular_velocity>-1</min_angular_velocity>
        </plugin>
    </gazebo>
</robot>